File: GFG/Python/Medium/kadanes-algorithm.py
# https://www.geeksforgeeks.org/problems/kadanes-algorithm-1587115620/1


class Solution:
    # #Complete this function
    # Function to find the sum of contiguous subarray with maximum sum.
    def max_sub_array_sum(self, arr):
        # #Your code here
        max_sum = float('-inf')
        curr_sum = 0

        for i in arr:
            if curr_sum < 0:
                curr_sum = 0

            curr_sum += i

            if curr_sum > max_sum:
                max_sum = curr_sum
        return max_sum

File: GFG/JavaScript/Medium/kadanes-algorithm.js
class Solution {
  // Function to find the sum of contiguous subarray with maximum sum.
  maxSubarraySum(arr) {
    // code here
    let max = -Infinity;
    let currMax = 0;

    for (let i = 0; i < arr.length; i++) {
      if (currMax < 0) {
        currMax = 0;
      }
      currMax += arr[i];
      if (max < currMax) {
        max = currMax;
      }
    }
    return max;
  }
}
